api/src/main/java/io/opentelemetry/common/ImmutableKeyValuePairs.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.common;
+
+import static io.opentelemetry.internal.Utils.checkArgument;
+import static io.opentelemetry.internal.Utils.checkNotNull;
+
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * An immutable set of key-value pairs. Keys are only {@link String} typed. Can be iterated over
+ * using the {@link #forEach(KeyValueConsumer)} method.
+ *
+ * @param <V> The type of the values contained in this.
+ * @see Labels
+ * @see Attributes
+ */
+@Immutable
+public interface ImmutableKeyValuePairs<V> {
+  /** Iterates over all the key-value pairs of attributes contained by this instance. */
+  void forEach(KeyValueConsumer<V> consumer);
+
+  /**
+   * Used for iterating over the key-value pairs contained by an {@link ImmutableKeyValuePairs}
+   * instance.
+   */
+  interface KeyValueConsumer<T> {
+    void consume(String key, T value);
+  }
+
+  class Helper {
+    private Helper() {}
+
+    static <T> List<Object> sortAndFilter(Object[] data) {
+      checkArgument(
+          data.length % 2 == 0, "You must provide an even number of key/value pair arguments.");
+
+      quickSort(data, 0, data.length - 2);
+      return dedupe(data);
+    }
+
+    private static void quickSort(Object[] data, int leftIndex, int rightIndex) {
```

Comment:
Guessing you wrote this you don't want to use `Arrays.sort()`, which uses merge sort (and more memory)? If that's the case, adding a note for further reference would be nice.

Reply:
I don't know of a way to use Arrays.sort() to sort only based on every other value, and keep the pairs together.

Reply:
Oooh, right. I totally forgot about that part ;)